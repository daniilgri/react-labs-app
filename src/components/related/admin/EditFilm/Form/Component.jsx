import React, { useEffect } from "react";
import { withRouter } from "react-router-dom";
import PropTypes from "prop-types";
import { useFormik } from "formik";

import DescriptionSection from "./DescriptionSection";
import MediaSection from "./MediaSection";
import FormalSection from "./FormalSection";

import { Wrapper, Head, Title, Body, FilledButton, Loading, RedirectLink } from "./styles";

const Component = ({
  openAddScreeningDateModal,
  filmId,
  film,
  editFilmRequested,
  fetchEditFilmRequested,
  loading,
  error,
}) => {
  useEffect(() => {
    fetchEditFilmRequested(filmId);
  }, [fetchEditFilmRequested, filmId]);

  const formik = useFormik({
    initialValues: {
      title: film.title,
      description: film.description,
      ticketPrice: film.ticketPrice,
      imageAsFile: film.image,
      previewImage: film.image,
      tags: film.tags,
      screeningDates: film.screeningDates,
    },
    onSubmit: values => {
      const changedValues = {};
      Object.keys(values).forEach(prop => {
        if (formik.initialValues[prop] !== values[prop]) {
          changedValues[prop] = values[prop];
        }
      });
      if (Object.keys(changedValues).length !== 0 && changedValues.constructor === Object) {
        editFilmRequested({ values, changedValues, filmId: film.id });
      }
    },
    enableReinitialize: true,
  });

  const handleImageUpload = event => {
    event.preventDefault();
    formik.setFieldValue("imageAsFile", event.currentTarget.files[0]);
  };

  const handleScreeningDateAddition = screeningDate => {
    formik.setFieldValue("screeningDates", [...formik.values.screeningDates, screeningDate]);
  };

  const handleScreeningDateDelete = screeningDate => {
    formik.setFieldValue(
      "screeningDates",
      formik.values.screeningDates.filter(sd => sd.date !== screeningDate.date)
    );
  };

  const handleTagAddition = tag => {
    formik.setFieldValue("tags", [...formik.values.tags, tag]);
  };

  const handleTagDelete = tag => {
    formik.setFieldValue(
      "tags",
      formik.values.tags.filter(elTag => elTag !== tag)
    );
  };

  const handleFormOnKeyDown = event => {
    if (event.key === "Enter") {
      event.preventDefault();
    }
  };

  if (loading || error) {
    return <Loading>Loading...</Loading>;
  }

  return (
    <Wrapper onKeyDown={handleFormOnKeyDown} onSubmit={formik.handleSubmit}>
      <Head>
        <Title>Edit film - {film.title}</Title>{" "}
        <RedirectLink to={`/admin/film/${film.id}`}>Go to film card</RedirectLink>
      </Head>
      <Body>
        <DescriptionSection
          title={formik.values.title}
          description={formik.values.description}
          tags={formik.values.tags}
          errors={formik.errors}
          onChange={formik.handleChange}
          onTagAdd={handleTagAddition}
          onTagDelete={handleTagDelete}
        />
        <MediaSection
          errors={formik.errors}
          onImageUpload={handleImageUpload}
          previewImage={formik.values.previewImage}
        />
        <FormalSection
          screeningDates={formik.values.screeningDates}
          ticketPrice={formik.values.ticketPrice}
          errors={formik.errors}
          onChange={formik.handleChange}
          onAddScreeningDateModalOpen={openAddScreeningDateModal}
          onScreeningDateAdd={handleScreeningDateAddition}
          onScreeningDateDelete={handleScreeningDateDelete}
        />
      </Body>

      <FilledButton type="submit">Publish</FilledButton>
    </Wrapper>
  );
};

Component.defaultProps = {
  error: "",
  filmId: "",
};

Component.propTypes = {
  editFilmRequested: PropTypes.func.isRequired,
  openAddScreeningDateModal: PropTypes.func.isRequired,
  filmId: PropTypes.string,
  film: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.bool, PropTypes.number, PropTypes.string, PropTypes.array])
  ).isRequired,
  fetchEditFilmRequested: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.string,
};

export default withRouter(Component);
