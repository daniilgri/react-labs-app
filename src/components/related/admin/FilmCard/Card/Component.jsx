import React, { useEffect } from "react";
import PropTypes from "prop-types";

import FilmBox from "./FilmBox";
import TicketBox from "./TicketBox";
import { Wrapper, Pending } from "./styles";

const Component = ({ filmId, fetchFilmByIdAdminPanelRequested, film, filmError, loading }) => {
  useEffect(() => {
    fetchFilmByIdAdminPanelRequested(filmId);
  }, [fetchFilmByIdAdminPanelRequested, filmId]);

  if (loading || filmError || !film) {
    return <Pending>Loading</Pending>;
  }

  return (
    <Wrapper>
      <FilmBox film={film} />
      <TicketBox screeningDates={film.screeningDates} />
    </Wrapper>
  );
};

Component.defaultProps = {
  filmError: "",
  filmId: "",
  film: {
    title: "",
    description: "",
    tags: [],
    image: "",
    rates: [],
    screeningDates: [],
    ticketPrice: "",
  },
};

Component.propTypes = {
  filmId: PropTypes.string,
  fetchFilmByIdAdminPanelRequested: PropTypes.func.isRequired,
  film: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.bool, PropTypes.number, PropTypes.string, PropTypes.array])
  ),
  filmError: PropTypes.string,
  loading: PropTypes.bool.isRequired,
};

export default Component;
