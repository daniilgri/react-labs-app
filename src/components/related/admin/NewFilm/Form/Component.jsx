import React from "react";
import { withRouter } from "react-router-dom";
import PropTypes from "prop-types";
import { useFormik } from "formik";

import DescriptionSection from "./DescriptionSection";
import MediaSection from "./MediaSection";
import FormalSection from "./FormalSection";

import newFilmSchema from "../../../../../validations/newFilmSchema";
import { Wrapper, Head, Title, Body, FilledButton } from "./styles";

const Component = ({ addFilmRequested, history, openAddScreeningDateModal }) => {
  const formik = useFormik({
    initialValues: {
      title: "",
      description: "",
      ticketPrice: 0,
      imageAsFile: "",
      tags: [],
      screeningDates: [],
    },
    validationSchema: newFilmSchema,
    onSubmit: values => {
      addFilmRequested(values);
      history.push("/");
    },
  });

  const handleImageUpload = event => {
    event.preventDefault();
    formik.setFieldValue("imageAsFile", event.currentTarget.files[0]);
  };

  const handleScreeningDateAddition = screeningDate => {
    formik.setFieldValue("screeningDates", [...formik.values.screeningDates, screeningDate]);
  };

  const handleScreeningDateDelete = screeningDate => {
    formik.setFieldValue(
      "screeningDates",
      formik.values.screeningDates.filter(sd => sd.date !== screeningDate.date)
    );
  };

  const handleTagAddition = tag => {
    formik.setFieldValue("tags", [...formik.values.tags, tag]);
  };

  const handleTagDelete = tag => {
    formik.setFieldValue(
      "tags",
      formik.values.tags.filter(elTag => elTag !== tag)
    );
  };

  const handleFormOnKeyDown = event => {
    if (event.key === "Enter") {
      event.preventDefault();
    }
  };

  return (
    <Wrapper onKeyDown={handleFormOnKeyDown} onSubmit={formik.handleSubmit}>
      <Head>
        <Title>Add new film</Title>
      </Head>
      <Body>
        <DescriptionSection
          values={formik.values}
          errors={formik.errors}
          onChange={formik.handleChange}
          onTagAdd={handleTagAddition}
          onTagDelete={handleTagDelete}
        />
        <MediaSection errors={formik.errors} onImageUpload={handleImageUpload} />
        <FormalSection
          values={formik.values}
          errors={formik.errors}
          onChange={formik.handleChange}
          onAddScreeningDateModalOpen={openAddScreeningDateModal}
          onScreeningDateAdd={handleScreeningDateAddition}
          onScreeningDateDelete={handleScreeningDateDelete}
        />
      </Body>

      <FilledButton type="submit">Publish</FilledButton>
    </Wrapper>
  );
};

Component.propTypes = {
  addFilmRequested: PropTypes.func.isRequired,
  history: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.func, PropTypes.number, PropTypes.string, PropTypes.object])
  ).isRequired,
  openAddScreeningDateModal: PropTypes.func.isRequired,
};

export default withRouter(Component);
