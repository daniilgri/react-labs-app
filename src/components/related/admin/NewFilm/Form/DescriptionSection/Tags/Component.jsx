import React, { useState } from "react";
import PropTypes from "prop-types";
import { v4 as uuidv4 } from "uuid";

import { Wrapper, List, Item, TagButton, Input, ErrorText } from "./styles";

const Component = ({ onSet, onDelete, values, errors }) => {
  const [tagInput, setTagInput] = useState("");

  const handleInputOnKeyDown = event => {
    if (event.key === "Enter") {
      if (!values.includes(tagInput) && !(tagInput === "")) {
        onSet(tagInput);
        setTagInput("");
      }
    }
  };

  const handleDeleteTagButtonOnClick = tag => () => {
    onDelete(tag);
  };

  const handleInputOnChange = event => {
    setTagInput(event.target.value);
  };

  return (
    <Wrapper>
      <List>
        {values.length > 0
          ? values.map(el => (
              // eslint-disable-next-line react/jsx-indent
              <Item key={uuidv4()}>
                {el}
                <TagButton type="button" onClick={handleDeleteTagButtonOnClick(el)}>
                  -
                </TagButton>
              </Item>
            ))
          : null}
      </List>
      <Input
        onKeyDown={handleInputOnKeyDown}
        placeholder="Type tag"
        name="tagInput"
        id="tagInput"
        value={tagInput}
        onChange={handleInputOnChange}
      />
      {errors.tags && <ErrorText>{errors.tags}</ErrorText>}
    </Wrapper>
  );
};

Component.propTypes = {
  onSet: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  values: PropTypes.oneOfType([PropTypes.object, PropTypes.array]).isRequired,
  errors: PropTypes.oneOfType([PropTypes.object, PropTypes.array]).isRequired,
};

export default Component;
