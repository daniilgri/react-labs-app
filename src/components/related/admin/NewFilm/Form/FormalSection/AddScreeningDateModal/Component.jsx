import React, { useState } from "react";
import PropTypes from "prop-types";
import { useFormik } from "formik";
import { v4 as uuidv4 } from "uuid";

import addScreeningDateSchema from "../../../../../../../validations/addScreeningDateSchema";
import {
  Wrapper,
  Content,
  Title,
  DateInput,
  TimeSection,
  FilledButton,
  ScreeningTime,
  Field,
  Label,
  SaveButton,
  ScreeningTimeInput,
  ScreeningTimeInputWrapper,
  ConfirmTimeAddButton,
  ScreeningTimeEditButtons,
} from "./styles";

const Component = ({ closeModal, isOpen, onScreeningDateAdd }) => {
  const [screeningTimeInputActive, setScreeningTimeInputActive] = useState(
    false
  );
  const [screeningTimeValue, setScreeningTimeValue] = useState("");

  const formik = useFormik({
    initialValues: {
      date: "",
      times: [],
    },
    validationSchema: addScreeningDateSchema,
    onSubmit: (values, { resetForm }) => {
      onScreeningDateAdd(values);

      setScreeningTimeInputActive(false);
      setScreeningTimeValue("");
      resetForm();
      closeModal();
    },
  });

  const handleScreeningTimeAddition = (event) => {
    event.preventDefault();

    formik.setFieldValue("times", [screeningTimeValue, ...formik.values.times]);

    setScreeningTimeValue("");
    setScreeningTimeInputActive(false);
  };

  const handleScreeningTimeCancel = (event) => {
    event.preventDefault();

    setScreeningTimeValue("");
    setScreeningTimeInputActive(false);
  };

  const handleAddTimeButtonOnClick = (event) => {
    event.preventDefault();

    setScreeningTimeValue("");
    setScreeningTimeInputActive((prevState) => !prevState);
  };

  const handleScreeningTimeOnChange = (event) => {
    setScreeningTimeValue(event.target.value);
  };

  const handleScreeningTimeItemDelete = (sdTime) => (event) => {
    event.preventDefault();

    formik.setFieldValue(
      "times",
      formik.values.times.filter((el) => el !== sdTime)
    );
  };

  return (
    <Wrapper
      isOpen={isOpen}
      onBackgroundClick={closeModal}
      onEscapeKeydown={closeModal}
    >
      <Content onSubmit={formik.handleSubmit}>
        <Title>Add screening date</Title>
        <Field>
          <Label>Date:</Label>
          <DateInput
            id="date"
            name="date"
            placeholder="date"
            onChange={formik.handleChange}
            value={formik.values.date}
          />
        </Field>
        <FilledButton type="button" onClick={handleAddTimeButtonOnClick}>
          Add time
        </FilledButton>
        <TimeSection>
          {screeningTimeInputActive && (
            <ScreeningTimeInputWrapper>
              <ScreeningTimeInput
                id="screeningTime"
                name="screeningTime"
                onChange={handleScreeningTimeOnChange}
                value={screeningTimeValue}
              />
              <ConfirmTimeAddButton onClick={handleScreeningTimeAddition}>
                +
              </ConfirmTimeAddButton>
              <ConfirmTimeAddButton onClick={handleScreeningTimeCancel}>
                -
              </ConfirmTimeAddButton>
            </ScreeningTimeInputWrapper>
          )}
          {formik.values.times.map((el) => (
            <ScreeningTime key={uuidv4()}>
              {el}
              <ScreeningTimeEditButtons>
                <ConfirmTimeAddButton
                  onClick={handleScreeningTimeItemDelete(el)}
                >
                  -
                </ConfirmTimeAddButton>
              </ScreeningTimeEditButtons>
            </ScreeningTime>
          ))}
        </TimeSection>
        <SaveButton type="submit">Save</SaveButton>
      </Content>
    </Wrapper>
  );
};

Component.defaultProps = {
  closeModal: () => {},
  isOpen: false,
  onScreeningDateAdd: () => {},
};

Component.propTypes = {
  closeModal: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onScreeningDateAdd: PropTypes.func.isRequired,
};

export default Component;
