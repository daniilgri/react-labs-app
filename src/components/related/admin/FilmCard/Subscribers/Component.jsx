import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { v4 as uuidv4 } from "uuid";
import Card from "./Card";

import { Wrapper, Loading, CenterContainer, FetchButton } from "./styles";

const Component = ({
  filmId,
  fetchSubscribersInitialRequested,
  users,
  loading,
  error,
  allCount,
  count,
  fetchSubscribersNextRequested,
}) => {
  const handleGetMoreButtonOnClick = event => {
    event.preventDefault();

    if (count < allCount) {
      fetchSubscribersNextRequested({ filmId });
    }
  };

  useEffect(() => {
    fetchSubscribersInitialRequested({ filmId });
  }, []);

  return (
    <>
      <Wrapper>
        {users.length > 0 && users.map(item => <Card key={uuidv4()} user={item} />)}
      </Wrapper>
      {loading || error ? (
        <Loading>Loading</Loading>
      ) : (
        count < allCount && (
          <CenterContainer>
            <FetchButton onClick={handleGetMoreButtonOnClick}>Get more</FetchButton>
          </CenterContainer>
        )
      )}
    </>
  );
};

Component.defaultProps = {
  users: [],
  error: "",
  count: 0,
  allCount: 0,
  filmId: "",
};

Component.propTypes = {
  filmId: PropTypes.string,
  fetchSubscribersInitialRequested: PropTypes.func.isRequired,
  users: PropTypes.arrayOf(PropTypes.object),
  error: PropTypes.string,
  loading: PropTypes.bool.isRequired,
  count: PropTypes.number,
  allCount: PropTypes.number,
  fetchSubscribersNextRequested: PropTypes.func.isRequired,
};

export default Component;
