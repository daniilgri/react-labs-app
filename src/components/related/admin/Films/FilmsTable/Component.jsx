import React, { useEffect } from "react";
import PropTypes from "prop-types";
import Card from "./Card";
import { v4 as uuidv4 } from "uuid";

import { Wrapper, Loading } from "./styles";

const Component = ({
  loading,
  fetchFilmsAdminPanelInitialRequested,
  films,
  error,
  deleteFilmRequested,
}) => {
  useEffect(() => {
    fetchFilmsAdminPanelInitialRequested({ count: 25 });
  }, []);

  let content;

  if (loading || error) {
    content = <Loading>Loading</Loading>;
  } else {
    content =
      films.length > 0 ? (
        films.map((item) => (
          <Card key={uuidv4()} film={item} onDelete={deleteFilmRequested} />
        ))
      ) : (
        <Loading>Empty</Loading>
      );
  }

  return <Wrapper>{content}</Wrapper>;
};

Component.defaultProps = {
  films: [],
  loading: true,
  error: "",
  fetchFilmsAdminPanelInitialRequested: () => {},
  deleteFilmRequested: () => {},
};

Component.propTypes = {
  films: PropTypes.array,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.string,
  fetchFilmsAdminPanelInitialRequested: PropTypes.func.isRequired,
  deleteFilmRequested: PropTypes.func.isRequired,
};

export default Component;
