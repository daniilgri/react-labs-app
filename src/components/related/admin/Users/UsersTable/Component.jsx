import React from "react";
import Card from "./Card";
import { v4 as uuidv4 } from "uuid";

import { Wrapper } from "./styles";

const Component = () => {
  const users = React.useState([1, 2, 3, 4, 5]);

  return (
    <Wrapper>
      {users.map((user) => (
        <Card key={uuidv4()} value={user} />
      ))}
    </Wrapper>
  );
};

export default Component;
