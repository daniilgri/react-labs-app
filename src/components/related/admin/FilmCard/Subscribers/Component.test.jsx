import React from "react";
import renderer from "react-test-renderer";

import Component from "./Component";

describe("Subscribers component", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Component
          loading={false}
          fetchSubscribersInitialRequested={() => {}}
          fetchSubscribersNextRequested={() => {}}
          filmId="filmId"
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
