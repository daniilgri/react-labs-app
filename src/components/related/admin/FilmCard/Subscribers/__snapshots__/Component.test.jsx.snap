// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscribers component renders correctly 1`] = `
<div
  className="sc-fubCfw hFaLrq"
/>
`;
